monitor/monitor.go

```diff
@@ -198,45 +270,74 @@ type readWriter struct {
 }
 
 type monitor struct {
-	muxIO           *ioset.MuxIO
-	invokeIO        *ioset.Forwarder
-	invokeFunc      func(context.Context, string, io.ReadCloser, io.WriteCloser, io.WriteCloser) error
-	curInvokeCancel func()
+	muxIO         *ioset.MuxIO
+	invokeIO      *ioset.Forwarder
+	invokeFunc    func(ctx context.Context, ref, pid string, cfg controllerapi.InvokeConfig, in io.ReadCloser, out io.WriteCloser, err io.WriteCloser) error
+	invokeCancel  func()
+	attachedPid   string
+	attachedPidMu sync.Mutex
+}
+
+func (m *monitor) rollback(ctx context.Context, ref string, cfg controllerapi.InvokeConfig) string {
+	pid := identity.NewID()
+	cfg1 := cfg
+	cfg1.Rollback = true
+	return m.startInvoke(ctx, ref, pid, cfg1)
+}
+
+func (m *monitor) exec(ctx context.Context, ref string, cfg controllerapi.InvokeConfig) string {
+	return m.startInvoke(ctx, ref, identity.NewID(), cfg)
+}
+
+func (m *monitor) attach(ctx context.Context, ref, pid string) {
+	m.startInvoke(ctx, ref, pid, controllerapi.InvokeConfig{})
 }
 
-func (m *monitor) rollback(ctx context.Context, ref string) {
-	if m.curInvokeCancel != nil {
-		m.curInvokeCancel() // Finish the running container if exists
+func (m *monitor) startInvoke(ctx context.Context, ref, pid string, cfg controllerapi.InvokeConfig) string {
+	if m.invokeCancel != nil {
+		m.invokeCancel() // Finish existing attach
 	}
 	go func() {
-		// Start a new container
-		if err := m.invoke(ctx, ref); err != nil {
+		// Start a new invoke
+		if err := m.invoke(ctx, ref, pid, cfg); err != nil {
 			logrus.Debugf("invoke error: %v", err)
 		}
+		m.attachedPidMu.Lock()
+		if pid == m.attachedPid {
+			m.attachedPid = ""
+		}
+		m.attachedPidMu.Unlock()
```

Comment:
We can use https://pkg.go.dev/sync/atomic@master#Value here. There's a few other places in this PR that we should use the atomic primitives for where we would otherwise need a mutex to access a single value.

Reply:
Fixed.